affinity/cpp-20/dXXX1r0.md

```diff
@@ -0,0 +1,373 @@
+
+Operating systems (OSes) traditionally take responsibility for assigning threads or processes to run on processing units. However, OSes may use high-level policies for this assignment that do not necessarily match the optimal usage pattern for a given application. Application developers must leverage the placement of memory and *placement of threads* for best performance on current and future architectures. For C\+\+ developers to achieve this, native support for *placement of threads and memory* is critical for application portability. We will refer to this as the *affinity problem*. 
+
+The affinity problem is especially challenging for applications whose behavior changes over time or is hard to predict, or when different applications interfere with each other's performance. Today, most OSes already can group processing units according to their locality and distribute processes, while keeping threads close to the initial thread, or even avoid migrating threads and maintain first touch policy. Nevertheless, most programs can change their work distribution, especially in the presence of nested parallelism.
+
+Frequently, data are initialized at the beginning of the program by the initial thread and are used by multiple threads. While some OSes automatically migrate threads or data for better affinity, migration may have high overhead. In an optimal case, the OS may automatically detect which thread access which data most frequently, or it may replicate data which are read by multiple threads, or migrate data which are modified and used by threads residing on remote locality groups. However, the OS often does a reasonable job, if the machine is not overloaded, if the application carefully uses first-touch allocation, and if the program does not change its behavior with respect to locality.
+
+Consider a code example *(Listing 1)* that uses the C\+\+17 parallel STL algorithm `for_each` to modify the entries of a `valarray` `a`.  The example applies a loop body in a lambda to each entry of the `valarray` `a`, using an execution policy that distributes work in parallel across multiple CPU cores. We might expect this to be fast, but since `valarray` containers are initialized automatically and automatically allocated on the master thread's memory, we find that it is actually quite slow even when we have more than one thread.
+
+```cpp
+// C++ valarray STL containers are initialized automatically.
+// First-touch allocation thus places all of a on the master.
+std::valarray<double> a(N);
+
+// Data placement is wrong, so parallel update is slow.
+std::for_each(std::execution::par, std::begin(a), std::end(a), 
+              [=] (double& a_i) { a_i *= scalar; });
+	      
+// Use future affinity interface to migrate data at next
```

Comment:
Per the call today (Mon 14 Jan 2019), we suggested keeping this example (perhaps changing valarray to vector), and showing the interface that would migrate pages of an existing allocation.  We also suggested adding an example that shows coupled control of allocation and execution -- e.g., "scatter."

Reply:
Thanks for the suggestion, since this PR is simply re-structuring I will merge this PR and open another with the new examples based on the discussion on today's call.